test/integration/protocol_integration_test.cc

```diff
@@ -4921,4 +4921,64 @@ TEST_P(DownstreamProtocolIntegrationTest, UnknownPseudoHeader) {
   }
 }
 
+// This test performs DNS refreshing very fast also keep changing DNS results.
+// It is trying to reproduce a R/W race for logical_dns_cluster.
+class LogicalDnsReadWriteRaceTest : public ProtocolIntegrationTest {
+public:
+  LogicalDnsReadWriteRaceTest() : registered_dns_factory_(dns_resolver_factory_) {}
+
+  NiceMock<Network::MockDnsResolverFactory> dns_resolver_factory_;
+  Registry::InjectFactory<Network::DnsResolverFactory> registered_dns_factory_;
+};
+
+INSTANTIATE_TEST_SUITE_P(
+    Protocols, LogicalDnsReadWriteRaceTest,
+    testing::ValuesIn(HttpProtocolIntegrationTest::getProtocolTestParamsWithoutHTTP3()),
+    HttpProtocolIntegrationTest::protocolTestParamsToString);
+
+TEST_P(LogicalDnsReadWriteRaceTest, FastChangingDnsResult) {
+  if (version_ == Network::Address::IpVersion::v4) {
+    return;
```

Comment:
i'm a little confused by this early-exit. You are only going to do this test in ipv6 mode, but it looks like you are mocking some ipv4 addresses below.

Reply:
We can reproduce the issue with either IPv4 or IPv6. So, we just do it with IPv6 here, to align with the above first returned address "::1".

If we want to support IPv4 here, the above DnsResponse has to be changed to "127.0.0.1" for IPv4, and "::1" for IPv6. We don't need that complexity.

Reply:
But you have ipv4 addresses below, so it seems inconsistent and confusing. If this is what you really intended can you write a comment explaining what you are trying to do?

Reply:
Yeah, I am trying to feed the HappyEyeball routine with both IPv4 and Ipv6 addresses to test it's sortAddresses() function.
At the end, it's the first address which will be used to make the connections. That's the reason the IP version of the test has to match with the first address type in the DnsResponse.  
Sure, let me add some comment to explain the intent of this test.